src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs

```diff
@@ -259,15 +259,17 @@ private BoundExpression VisitIndexPatternIndexerAccess(BoundImplicitIndexerAcces
             switch (strategy)
             {
                 case PatternIndexOffsetLoweringStrategy.SubtractFromLength:
-                    // ensure we evaluate the input before accessing length
-                    if (makeOffsetInput.ConstantValue is null)
+                    BoundExpression lengthAccess = RewriteLengthAccess(node, receiver);
+
+                    // ensure we evaluate the input before accessing length, unless it is an array length
```

Comment:
why is array length different in terms of ordering?

Reply:
> why is array length different in terms of ordering?

Because it is an IL instruction, not a property access. I.e. not observable.